### PR TITLE
#2287 Add example scripts showing pub/sub usage.

### PR DESCRIPTION
**Pull request by con-mark**
### Description

Added example requested for Issue #2287 

created pubsub-publisher.js and pubsub-subscriber.js

---

### Checklist

- [ ] Does `npm test` pass with this change (including linting)?
- [ ] Is the new or changed code fully tested?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?

### CODE REVIEW

**Comment by SuzeShardlow** [comment]
Closes #2287 when approved and merged.

**Comment by SuzeShardlow** [comment]
@con-mark thanks for your PR - not sure if you have seen @simonprickett's comments above, but he is asking for a couple of changes before we merge this.  Thanks!

**Comment by con-mark** [comment]
Hi @simonprickett  @SuzeShardlow  I have made requested changes please let me know if any further changes are needed.

**Comment by SuzeShardlow** [comment]
Thanks @con-mark, we will take a look in our morning! :)

File: examples/pubsub-publisher.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+await client.connect();
+
+//declare constant variables for the name of the clients we will publish to as they will be required for logging
+const channel1 = 'chan1nel';
+const channel2 = 'chan2nel';
+
+for (let i = 0; i < 10000; i++) {
+  //1st channel created to publish 10000 messages
+  await client.publish(channel1, `channel1_message_${i}`);
+  console.log(`publising message on ${channel1}`);
```
Comment (simonprickett) [inline examples/pubsub-publisher.js:16]:
```suggestion
  console.log(`publishing message on ${channel1}`);

```

File: examples/pubsub-publisher.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  //2nd channel created to publish 10000 messages
+  await client.publish(channel2, `channel2_message_${i}`);
+  console.log(`publising message on ${channel2}`);
```
Comment (simonprickett) [inline examples/pubsub-publisher.js:19]:
```suggestion
  console.log(`publishing message on ${channel2}`);

```

File: examples/pubsub-publisher.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+// A sample publisher using the publish function to put message on different channels.
+// https://redis.io/commands/publish/
+import { createClient } from 'redis';
+
+const client = createClient();
+
+await client.connect();
+
+//declare constant variables for the name of the clients we will publish to as they will be required for logging
```
Comment (simonprickett) [inline examples/pubsub-publisher.js:9]:
```suggestion
// Declare constant variables for the name of the clients we will publish to as they will be required for logging.

```

File: examples/pubsub-publisher.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+const client = createClient();
+
+await client.connect();
+
+//declare constant variables for the name of the clients we will publish to as they will be required for logging
+const channel1 = 'chan1nel';
+const channel2 = 'chan2nel';
+
+for (let i = 0; i < 10000; i++) {
+  //1st channel created to publish 10000 messages
```
Comment (simonprickett) [inline examples/pubsub-publisher.js:14]:
```suggestion
  // 1st channel created to publish 10000 messages.
```

File: examples/pubsub-publisher.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  await client.publish(channel1, `channel1_message_${i}`);
+  console.log(`publising message on ${channel1}`);
+  //2nd channel created to publish 10000 messages
```
Comment (simonprickett) [inline examples/pubsub-publisher.js:17]:
```suggestion
  // 2nd channel created to publish 10000 messages.
```

File: examples/pubsub-subscriber.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
+// A sample subscriber showing how the subscribe method and pSubscribe method work.
+// https://redis.io/commands/subscribe/
+// https://redis.io/commands/pSubscribe/
+// This consumes messages published by pubsub-publisher.js
+
+import { createClient} from 'redis';
+
+//create and connect client before doing any redis commands
```
Comment (simonprickett) [inline examples/pubsub-subscriber.js:8]:
```suggestion
// create and connect client before executing any Redis commands.
```

File: examples/pubsub-subscriber.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+// https://redis.io/commands/pSubscribe/
+// This consumes messages published by pubsub-publisher.js
+
+import { createClient} from 'redis';
+
+//create and connect client before doing any redis commands
+const client = createClient();
+await client.connect();
+
+//Each subscriber needs to connect individually therefore we duplicate the client.
```
Comment (simonprickett) [inline examples/pubsub-subscriber.js:12]:
```suggestion
// Each subscriber needs to connect individually therefore we duplicate the client.
```

File: examples/pubsub-subscriber.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+const channel2Sub = client.duplicate();
+const noChannelsSub = client.duplicate();
+const allChannelsSub = client.duplicate();
+
+await channel1Sub.connect();
+await channel2Sub.connect();
+await noChannelsSub.connect();
+await allChannelsSub.connect();
+
+//This subscriber only will receive messages from channel 1 as they are using the subscribe method and subscribed to chan1nel
```
Comment (simonprickett) [inline examples/pubsub-subscriber.js:23]:
```suggestion
// This subscriber only will receive messages from channel 1 as they are using the subscribe method and subscribed to chan1nel.

```

File: examples/pubsub-subscriber.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+await channel1Sub.subscribe('chan1nel', (message) => {
+  console.log(`Channel1 subscriber collected message: ${message}`);
+},true);
+
+//This subscriber only will receive messages from channel 2 as they are using the subscribe method and subscribed to chan2nel
```
Comment (simonprickett) [inline examples/pubsub-subscriber.js:28]:
```suggestion
// This subscriber only will receive messages from channel 2 as they are using the subscribe method and subscribed to chan2nel.
```

File: examples/pubsub-subscriber.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+await channel2Sub.subscribe('chan2nel', (message) => {
+  console.log(`Channel2 subscriber collected message: ${message}`);
+},true);
+
+//This subscriber not receive any messages cause its channel does not exist
+await noChannelsSub.subscribe('chan*nel', (message) => {
+  console.log(`Channel2 subscriber collected message: ${message}`);
+},true);
+
+//This subscriber receive messages from both channel 1 and channel 2 thanks the pSubscribe method.
```
Comment (simonprickett) [inline examples/pubsub-subscriber.js:38]:
```suggestion
// This subscriber receive messages from both channel 1 and channel 2 using the pSubscribe method.
```

File: examples/pubsub-subscriber.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+await allChannelsSub.pSubscribe('chan*nel', (message, channel) => {
+  console.log(`Channel ${channel} sent message: ${message}`);
+},true);
+
+//clean up after ourselves by unsubsribing
```
Comment (simonprickett) [inline examples/pubsub-subscriber.js:43]:
```suggestion
// Clean up after ourselves by unsubscribing.

```

**Review by simonprickett — CHANGES_REQUESTED** [review]
I've left some tidy up commit suggestions, but am also unable to get the example to work, the publisher publishes messages but the subscriber never outputs anything...

**Comment by SuzeShardlow** [comment]
@con-mark, please see the above comments from @simonprickett and let us know if you can make the changes - the subscriber in your example isn't outputting anything for us.  Thanks!

**Comment by con-mark** [comment]
Hi @SuzeShardlow  and @simonprickett  I have made the tidy up comments and I have fixed the issue with the subscriber seems like doing the clean up and closing the client before they got a chance to interact with the Redis to publish or subscribe.

Please let me know if you need anything more.

I also may suggest some other improvement for your hacktoberfest contribution might be for someone to create a docker compose that will allow the the running of the examples with a redis image for people to not have to worry for set up :)

File: examples/pubsub-subscriber.js (hunk 11 above)
Comment (con-mark) [reply examples/pubsub-subscriber.js]: removed clean up

File: examples/pubsub-subscriber.js (hunk 10 above)
Comment (con-mark) [reply examples/pubsub-subscriber.js]: done

File: examples/pubsub-subscriber.js (hunk 9 above)
Comment (con-mark) [reply examples/pubsub-subscriber.js]: done

File: examples/pubsub-subscriber.js (hunk 8 above)
Comment (con-mark) [reply examples/pubsub-subscriber.js]: done

File: examples/pubsub-subscriber.js (hunk 7 above)
Comment (con-mark) [reply examples/pubsub-subscriber.js]: done

File: examples/pubsub-publisher.js (hunk 5 above)
Comment (con-mark) [reply examples/pubsub-publisher.js]: done

File: examples/pubsub-publisher.js (hunk 4 above)
Comment (con-mark) [reply examples/pubsub-publisher.js]: done

File: examples/pubsub-publisher.js (hunk 3 above)
Comment (con-mark) [reply examples/pubsub-publisher.js]: done

File: examples/pubsub-publisher.js (hunk 2 above)
Comment (con-mark) [reply examples/pubsub-publisher.js]: done

File: examples/pubsub-publisher.js (hunk 1 above)
Comment (con-mark) [reply examples/pubsub-publisher.js]: done

File: examples/pubsub-subscriber.js [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+await channel1Sub.subscribe('chan1nel', (message) => {
+  console.log(`Channel1 subscriber collected message: ${message}`);
+},true);
+
+// This subscriber only will receive messages from channel 2 as they are using the subscribe method and subscribed to chan2nel.
+await channel2Sub.subscribe('chan2nel', (message) => {
+  console.log(`Channel2 subscriber collected message: ${message}`);
+},true);
+
+//This subscriber not receive any messages cause its channel does not exist
```
Comment (simonprickett) [inline examples/pubsub-subscriber.js:33]:
```suggestion
// This subscriber will not receive any messages as its channel does not exist.
```

**Review by simonprickett — CHANGES_REQUESTED** [review]
@con-mark thanks - left one tiny suggestion and this should then be good to go.

**Comment by con-mark** [comment]
Hi @SuzeShardlow  @simonprickett  is any further changes needed here? Please let me know :)

**Review by simonprickett — APPROVED** [review]

**Comment by SuzeShardlow** [comment]
@con-mark thanks for your contribution!  If you are a member of our Discord server, please let us know your handle so we can give you the `Hacktoberfest 2022` role.  Thanks!

**Comment by con-mark** [comment]
Hi @SuzeShardlow my discord is dev-mark#1608

**Comment by SuzeShardlow** [comment]
Thanks @con-mark, I've given you the role in Discord :)

**Pull request merged**